Tests/WeaverDITests/ThreadSafeBuilderStoringDecoratorTests.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  ThreadSafeBuilderStoringDecoratorTests.swift
+//  WeaverDITests
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+import XCTest
+
+@testable import WeaverDI
+
+class ThreadSafeBuilderStoringDecoratorTests: XCTestCase {
+    
+
+    let instanceKey = InstanceKey(for: String.self, name: "test")
+    
+    func builder() -> String {
+        return "builder"
+    }
+    
+    func test_set_should_delegate_to_inner_builder() {
+        let builderMock = BuilderStoreMock()
+        
+        let sut = ThreadSafeBuilderStoreDecorator(builders: builderMock)
+        sut.set(builder: builder, scope: .container, for: instanceKey)
+        
+        waitDelay()
+        
+        XCTAssertEqual(builderMock.setCallCount, 1)
+        XCTAssertEqual((builderMock.builder as? ()->String)?(), "builder")
+        XCTAssertEqual(builderMock.scope, .container)
+        XCTAssertEqual(builderMock.key, instanceKey)
+    }
+    
+   
+    func test_get_should_delegate_to_inner_builder() {
+        let builderMock = BuilderStoreMock()
+        
+        let sut = ThreadSafeBuilderStoreDecorator(builders: builderMock)
+        sut.set(builder: builder, scope: .container, for: instanceKey)
+        
+        let result: (()-> String)? = sut.get(for: instanceKey, isCalledFromAChild: true)
+                
+        XCTAssertEqual(builderMock.getCallCount, 1)
+        XCTAssertEqual(result?(), "builder")
+        XCTAssertEqual(builderMock.key, instanceKey)
+    }
+    
+    func test_many_access_in_parallel() {
```

Comment:
Let's write this tests the same way it's written in `DependencyContainerTests.swift`